### PR TITLE
Correct closest points box mesh

### LINKED ISSUE #66: Issue with closest_points_on_boundaries
### Describe the bug, what's wrong, and what you expected.

There is a bug in the function closest_points_on_boundaries

Considering a regular mesh on a 1,1,1 box with points on the corners, center of edges and center of faces such function should return a single closest neighbor for all the edges

### Steps to reproduce the bug.

```python

def _box_mesh_points() -> npt.NDArray[np.float_]:
    points = np.array(
        [
            [1.0, 1.0, 1.0],
            [1.0, 0.5, 1.0],
            [1.0, 0.0, 1.0],
            [1.0, 1.0, 0.5],
            [1.0, 0.5, 0.5],
            [1.0, 0.0, 0.5],
            [1.0, 1.0, 0.0],
            [1.0, 0.5, 0.0],
            [1.0, 0.0, 0.0],
            [0.5, 1.0, 1.0],
            [0.5, 0.5, 1.0],
            [0.5, 0.0, 1.0],
            [0.5, 1.0, 0.5],
            [0.5, 0.5, 0.5],
            [0.5, 0.0, 0.5],
            [0.5, 1.0, 0.0],
            [0.5, 0.5, 0.0],
            [0.5, 0.0, 0.0],
            [0.0, 1.0, 1.0],
            [0.0, 0.5, 1.0],
            [0.0, 0.0, 1.0],
            [0.0, 1.0, 0.5],
            [0.0, 0.5, 0.5],
            [0.0, 0.0, 0.5],
            [0.0, 1.0, 0.0],
            [0.0, 0.5, 0.0],
            [0.0, 0.0, 0.0],
        ]
    )
    return points


def _box_mesh_elements() -> Dict[pv.CellType, npt.NDArray[np.int_]]:
    elements_dict = {
        pv.CellType.TETRA: np.array(
            [
                [3, 4, 1, 9],
                [3, 1, 0, 9],
                [14, 16, 8, 17],
                [6, 4, 3, 12],
                [1, 9, 4, 10],
                [4, 5, 2, 10],
                [4, 2, 1, 10],
                [4, 9, 3, 12],
                [12, 10, 9, 18],
                [10, 12, 4, 13],
                [6, 7, 4, 12],
                [5, 11, 10, 13],
                [2, 10, 5, 11],
                [8, 5, 7, 13],
                [10, 9, 4, 12],
                [4, 12, 7, 13],
                [7, 5, 4, 13],
                [5, 10, 4, 13],
                [13, 12, 7, 15],
                [12, 6, 7, 15],
                [11, 13, 5, 14],
                [5, 13, 8, 14],
                [14, 13, 8, 16],
                [13, 7, 8, 16],
                [15, 7, 13, 16],
                [12, 13, 10, 18],
                [15, 13, 12, 21],
                [10, 18, 13, 19],
                [14, 11, 13, 19],
                [13, 11, 10, 19],
                [13, 18, 12, 21],
                [14, 20, 19, 22],
                [19, 21, 13, 22],
                [19, 14, 11, 20],
                [15, 16, 13, 21],
                [21, 15, 16, 24],
                [17, 14, 16, 22],
                [19, 18, 13, 21],
                [13, 21, 16, 22],
                [16, 14, 13, 22],
                [14, 19, 13, 22],
                [20, 22, 14, 23],
                [14, 22, 17, 23],
                [22, 21, 16, 24],
                [17, 22, 16, 25],
                [22, 24, 16, 25],
                [23, 22, 17, 25],
                [23, 25, 17, 26],
            ]
        )
    }

    return elements_dict


mesh = BoxMesh(_box_mesh_points(), _box_mesh_elements())
rve = Rve(dim=(1, 1, 1), center=(0.5, 0.5, 0.5))


mesh.nodes_coords[mesh.faces["face_xp"][0]] *= 1.01
print(mesh.nodes_coords[mesh.faces["face_xp"][0]])

clos = mesh.closest_points_on_boundaries(k_neighbours=3)

print(clos)
```

Here is the result:
{'face_xp': (array([[4, 3, 5]]), array([[0.01224745, 0.5       , 0.5       ]])), 'face_yp': (array([[12]]), array([[0.]])), 'face_zp': (array([[10]]), array([[0.]])), 'edge_xpym': ([[5]], [[0.0]]), 'edge_xpyp': ([[3]], [[0.0]]), 'edge_xmyp': ([[21]], [[0.0]]), 'edge_xpzm': ([[7]], [[0.0]]), 'edge_xpzp': ([[1]], [[0.0]]), 'edge_xmzp': ([[19]], [[0.0]]), 'edge_ypzm': ([[6, 24]], [[0.5, 0.5]]), 'edge_ypzp': ([[9]], [[0.0]]), 'edge_ymzp': ([[11]], [[0.0]])}

edge_ypzm has wrong results, should be 'edge_ypzm': ([[15]], [[0.0]])

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Fri May 17 17:05:28 2024 CEST

                OS : Darwin (macOS 14.0)
            CPU(s) : 12
           Machine : arm64
      Architecture : 64bit
       Environment : Python

  Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:49:36)
  [Clang 16.0.6 ]

           pyvista : 0.43.7
               vtk : 9.2.6
             numpy : 1.26.4
            scooby : 0.10.0
              gmsh : 4.13.0
          cadquery : 2.4.0
        matplotlib : 3.8.4
            meshio : 5.3.5
--------------------------------------------------------------------------------
```

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by chemiskyy**
This PR fixes #66. Tests are added for closest_points_on_boundaries

### CODE REVIEW

File: microgen/box_mesh.py
Comment (chemiskyy) [inline microgen/box_mesh.py:1]: The bug is on line 508

File: tests/test_box_mesh.py [hunk 1]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+
+def test_closest_points_on_perturbated_mesh_boundaries_must_have_good_number_of_neighbours() -> (
+    None
+):
+    box_mesh_to_test = BoxMesh(_box_mesh_points(), _box_mesh_elements())
+    dummy_noise_factor = 0.01
+
+    # perturb faces and edges with non-zero perturbation so that multiple neighbours must be found
+    for key, value in box_mesh_to_test.faces.items():
+        box_mesh_to_test.nodes_coords[value[0]] += dummy_noise_factor * (
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:255]: What is the meaning of `value[0]` ? 

Reading the code I think is the first element of the list of `faces` or `edges` ? when `box_mesh_to_test.faces` or `box_mesh_to_test.edges` is called ? Can you use a better variable name other than `value` for these two for loops ?

File: tests/test_box_mesh.py [hunk 2]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+            1 - random.uniform(0, 1)
+        )
+
+    for key, value in box_mesh_to_test.edges.items():
+        box_mesh_to_test.nodes_coords[value[0]] += dummy_noise_factor * (
+            1 - random.uniform(0, 1)
+        )
+
+    box_mesh_to_test.nodes_coords = np.clip(box_mesh_to_test.nodes_coords, 0.0, 1.0)
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:264]: Why we should clip the nodes coordinates between 0 and 1 ? If we change this clip by other arbitrary range the test should pass ?

File: tests/test_box_mesh.py [hunk 3]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+
+    closest_pts = box_mesh_to_test.closest_points_on_boundaries(k_neighbours=3)
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:266]: Is a particular reason to use `3` `k_neighbours` for this test ? Can you use a variable name to make it clear please ?

File: tests/test_box_mesh.py [hunk 4]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+
+    closest_neighbours_truth = [3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2]
+    closest_neighbours_to_test = []
+    for key, value in closest_pts.items():
+        closest_neighbours_to_test.append(len(value[0][0]))
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:271]: What represents `len(value[0][0])`

**Review by ricardo0115 — CHANGES_REQUESTED** [review]
Test logic LGTM a minor changes in variable naming may be required in order to facilitate the reading of this tests in future :) Thanks

File: tests/test_box_mesh.py [hunk 5]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
         bool_check_triangle_on_boundary_list
     )
+
+
+def test_closest_points_on_boundaries_must_have_good_number_of_neighbours() -> None:
+    box_mesh_to_test = BoxMesh(_box_mesh_points(), _box_mesh_elements())
+    closest_pts = box_mesh_to_test.closest_points_on_boundaries(k_neighbours=3)
+
+    for key, value in closest_pts.items():
+        assert (len(value[0][0])) == 1
```
[the remark this answers is not in the document]
Comment (chemiskyy) [reply tests/test_box_mesh.py:244]: explicit names have be given to value[0]. The extend method has been utilized to append all the lists of neighbors and avoid having to test only the first one (value[0][0])

File: tests/test_box_mesh.py [hunk 6]
```diff
@@ -232,3 +234,40 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+
+
+def test_closest_points_on_perturbated_mesh_boundaries_must_have_good_number_of_neighbours() -> (
+    None
+):
+    box_mesh_to_test = BoxMesh(_box_mesh_points(), _box_mesh_elements())
+    dummy_noise_factor = 0.01
```
[the remark this answers is not in the document]
Comment (chemiskyy) [reply tests/test_box_mesh.py:251]: perturbation_factor is utilized

File: tests/test_box_mesh.py (hunk 2 above)
Comment (chemiskyy) [reply tests/test_box_mesh.py:264]: The clip applies to keep all nodes in the original bounding box.

It has been replaced with
```
xyz_min_values_of_rve = np.min(mesh.rve.min_point)
xyz_max_values_of_rve = np.max(mesh.rve.max_point)

```
...
```
box_mesh_to_test.nodes_coords = np.clip(
        box_mesh_to_test.nodes_coords, xyz_min_values_of_rve, xyz_max_values_of_rve
    )
```

File: tests/test_box_mesh.py (hunk 3 above)
Comment (chemiskyy) [reply tests/test_box_mesh.py:266]: `@pytest.mark.parametrize("k_neighbours", [1, 2, 3, 4])` is now utilized

File: tests/test_box_mesh.py [hunk 7]
```diff
@@ -232,3 +234,86 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+    k_neighbours: int,
+) -> None:
+    box_mesh_to_test = BoxMesh(_box_mesh_points(), _box_mesh_elements())
+    perturbation_factor = 0.01
+
+    xyz_min_values_of_rve = np.min(box_mesh_to_test.rve.min_point)
+    xyz_max_values_of_rve = np.max(box_mesh_to_test.rve.max_point)
+
+    # perturb faces and edges with non-zero perturbation so that multiple neighbours must be found
+    for key, node_index_in_faces in box_mesh_to_test.faces.items():
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:275]:
```suggestion
    for _, node_index_in_faces in box_mesh_to_test.faces.items():

```

As the variable name "key" is not used in this loop you can ignored it using "_"

File: tests/test_box_mesh.py [hunk 8]
```diff
@@ -232,3 +234,86 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+        box_mesh_to_test.nodes_coords[node_index_in_faces[0]] += perturbation_factor * (
+            1 - random.uniform(0, 1)
+        )
+
+    for key, node_index_in_edges in box_mesh_to_test.edges.items():
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:280]:
```suggestion
    for _, node_index_in_edges in box_mesh_to_test.edges.items():

```

As the variable name "key" is not used in this loop you can ignored it using "_"

File: tests/test_box_mesh.py [hunk 9]
```diff
@@ -232,3 +234,86 @@ def test_given_box_box_mesh_boundary_elements_must_find_boundary_surface_element
...
+    k_neighbours: int,
+) -> None:
+    box_mesh_to_test = BoxMesh(_box_mesh_points(), _box_mesh_elements())
+    closest_pts = box_mesh_to_test.closest_points_on_boundaries(
+        k_neighbours=k_neighbours
+    )
+
+    number_of_closest_neighbours_to_test: List[int] = []
+    number_closest_neighbours_truth_1_on_each_face_and_edge = [1] * 12
+    for key, list_of_neighbor_for_all_nodes in closest_pts.items():
```
Comment (ricardo0115) [inline tests/test_box_mesh.py:250]:
```suggestion
    for _, list_of_neighbor_for_all_nodes in closest_pts.items():

```

**Review by ricardo0115 — APPROVED** [review]
LGTM ! Thanks !

**Pull request merged**